### PR TITLE
Adds warning when 'profile cascading' wont work, because of the order

### PR DESCRIPTION
**Pull request by welzfabian**
... in which profiles are applied.

```
14:01:18.926 WARN  [main|20ms] permissions - Profile 'test-cascade-to-target-with-equal-priority' refers to a profile wich is applied ealier than itself ('cascade-target'). Therefore the profiles will not be resolved completely. Fix this by adding priorities.
14:01:18.928 WARN  [main|22ms] permissions - Profile 'test-cascade-to-target-with-lower-priority' refers to a profile wich is applied ealier than itself ('cascade-target'). Therefore the profiles will not be resolved completely. Fix this by adding priorities.

### CODE REVIEW

File: src/main/java/sirius/web/security/Profile.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import sirius.kernel.settings.Extension;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Represents a profile defined in <tt>security.profiles</tt>.
+ */
+public class Profile {
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Profile.java]: does this need to be public?

File: src/main/java/sirius/web/security/Profile.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    public static final String SECURITY_PROFILES = "security.profiles";
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Profile.java]: javadoc

File: src/main/java/sirius/web/security/Profile.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    private final String name;
+    private final Set<String> permissionsToAdd;
+    private final Set<String> permissionsToRemove;
+
+    /**
+     * Create a Profile with the given name, permissions this profiles adds and permissions this profile removes.
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Profile.java]:
```suggestion
     * Creates a Profile with the given name, permissions this profiles adds and permissions this profile removes.
```

File: src/main/java/sirius/web/security/Profile.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @param name                the name of the profile
+     * @param permissionsToAdd    permissions this profile adds
+     * @param permissionsToRemove permission this profiles removes
+     */
+    public Profile(String name, Set<String> permissionsToAdd, Set<String> permissionsToRemove) {
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Profile.java]: make private as compile() should/must be called?

File: src/main/java/sirius/web/security/Profile.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        this.name = name;
+        this.permissionsToAdd = new HashSet<>(permissionsToAdd);
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Profile.java]: can be removed if private

File: src/main/java/sirius/web/security/Profile.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+     * Validates this profile and throws exception if problems exist.
+     * <p>
+     * An exception will be thrown if the profile refers to another profile applied earlier than itself.
+     */
+    public void validate() {
+        Extension thisProfile = Sirius.getSettings().getExtension(SECURITY_PROFILES, name);
+        for (String permission : thisProfile.getContext().keySet()) {
+            Extension otherProfile = Sirius.getSettings().getExtension(SECURITY_PROFILES, permission);
+            if (otherProfile == null || otherProfile.isDefault()) {
+                continue;
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Profile.java]: nope

File: src/main/java/sirius/web/security/Permissions.java [hunk 7]
```diff
@@ -94,26 +71,17 @@ private Permissions() {
...
+            profiles.add(profile);
+            try {
+                profile.validate();
+            } catch (Exception e) {
+                LOG.WARN(e.getMessage());
             }
         }
 
-        return profile;
+        profilesCache = profiles;
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Permissions.java]: return here and assign in getProfiles(?)

File: src/main/java/sirius/web/security/Profile.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        if (Permissions.hasPermission(name, permissions::contains)) {
+            permissions.addAll(permissionsToAdd);
+            permissions.removeAll(permissionsToRemove);
+        }
+    }
+
+    /**
+     * Validates this profile and throws exception if problems exist.
+     * <p>
+     * An exception will be thrown if the profile refers to another profile applied earlier than itself.
```
Comment (andyHa) [inline src/main/java/sirius/web/security/Profile.java]:
```suggestion
     * @throws HandledException if the profile refers to another profile applied earlier than itself.
```

File: src/main/java/sirius/web/security/Profile.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        if (Permissions.hasPermission(name, permissions::contains)) {
+            permissions.addAll(permissionsToAdd);
+            permissions.removeAll(permissionsToRemove);
+        }
+    }
+
+    /**
+     * Validates this profile and throws exception if problems exist.
+     * <p>
+     * @throws HandledException if the profile refers to another profile applied earlier than itself.
```
Comment (sabieber) [inline src/main/java/sirius/web/security/Profile.java]:
```suggestion
     * @throws IllegalStateException if the profile refers to another profile applied earlier than itself.
```

File: src/main/java/sirius/web/security/Profile.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+     * @throws IllegalStateException if the profile refers to another profile applied earlier than itself.
+     */
+    public void validate() {
+        Extension thisProfile = Sirius.getSettings().getExtension(SECURITY_PROFILES, name);
+        for (String permission : thisProfile.getContext().keySet()) {
+            Extension otherProfile = Sirius.getSettings().getExtension(SECURITY_PROFILES, permission);
+            if (otherProfile != null && !otherProfile.isDefault() && otherProfile.compareTo(thisProfile) <= 0) {
+                throw new IllegalStateException(Strings.apply(
+                        "Profile '%s' refers to a profile which is applied earlier than itself ('%s'). "
+                        + "Therefore the profiles will not be resolved completely. Fix this by adding priorities.",
```
Comment (meg1502) [inline src/main/java/sirius/web/security/Profile.java]:
```suggestion
                        + "Therefore the profiles will not be resolved completely. Fix this by adding or changing priorities.",
```
the problem can still occur if wrong priorities are added to the profiles, this should be reflected in the error message

File: src/test/java/sirius/web/security/ProfileTest.groovy [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+class ProfileTest extends BaseSpecification {
+
+    def "profile cascading is invalid when cascading to lower priority"() {
+        when:
+        Profile.compile(Sirius.getSettings().getExtension("security.profiles",
+                                                          "test-cascade-to-target-with-lower-priority"))
+               .validate()
+        then:
+        def e = thrown(IllegalStateException)
+        e.getMessage() == "Profile 'test-cascade-to-target-with-lower-priority' refers to a profile which is applied " +
```
Comment (meg1502) [inline src/test/java/sirius/web/security/ProfileTest.groovy]: can be exported into ```createErrorMessage(profile1,profile2)```to reduce test maintenace cost when editing this error message

**Pull request merged**